internal/controller/stages/stages.go

```diff
@@ -676,16 +683,40 @@ func (r *reconciler) syncNormalStage(
 			"Stage has no current Freight; no health checks or verification to perform",
 		)
 	} else {
-		// Always check the health of the Argo CD Applications associated with the
-		// Stage. This is regardless of the phase of the Stage, as the health of the
-		// Argo CD Applications is always relevant.
-		if status.Health = r.appHealth.EvaluateHealth(
-			ctx,
-			stage,
-		); status.Health != nil {
-			logger.WithValues("health", status.Health.Status).Debug("Stage health assessed")
+		if stage.Spec.PromotionTemplate != nil &&
+			stage.Status.LastPromotion != nil &&
+			stage.Status.LastPromotion.Status != nil {
+			if len(stage.Status.LastPromotion.Status.HealthChecks) > 0 {
+				var steps []directives.HealthCheckStep
+				for _, step := range stage.Status.LastPromotion.Status.HealthChecks {
+					steps = append(steps, directives.HealthCheckStep{
+						Kind:   step.Step,
+						Config: step.GetConfig(),
+					})
+				}
+
+				health := r.directivesEngine.CheckHealth(ctx, directives.HealthCheckContext{
+					Project: stage.Namespace,
+					Stage:   stage.Name,
+				}, steps)
+				status.Health = &health
+
+				logger.WithValues("health", status.Health.Status).Debug("Stage health assessed")
+			} else {
+				logger.Debug("Stage has no health checks to perform")
+			}
 		} else {
-			logger.Debug("Stage health deemed not applicable")
+			// Always check the health of the Argo CD Applications associated with the
+			// Stage. This is regardless of the phase of the Stage, as the health of the
+			// Argo CD Applications is always relevant.
+			if status.Health = r.appHealth.EvaluateHealth(
+				ctx,
+				stage,
+			); status.Health != nil {
+				logger.WithValues("health", status.Health.Status).Debug("Stage health assessed")
+			} else {
+				logger.Debug("Stage health deemed not applicable")
+			}
```

Comment:
It look like we're doing the classic health check as long as the promotion status has 0 healthchecks defined...

I don't feel 100% positive, but I am thinking this may not be the ideal behavior.

Someone could be using directives instead of classic promotion mechanisms and as long as they haven't updated any Argo CD Apps (since this is the only promotion step that currently generates any health checks steps) then they would have 0 health check steps in that collection, but falling back on the legacy health check process doesn't seem right in this case, especially since it is tightly coupled to the legacy promotion mechanisms, which are not at play in this case.

What would you think about only doing these legacy health checks if the legacy promo mechs are non-nil?

Reply:
If people would have legacy promotion mechanisms and new directives defined, that would also cause issues because then it could potentially never check the new health checks.

But I get your point, and will restructure the logic a bit and e.g. make use of a switch to avoid the if/else issue.